File: TypeScript/autotest_mypage/src/loadable-pages/login-page.ts
import { Locator, Page } from "@playwright/test";
import { BasePage } from "../base-page";
import { InputWebElement } from "../element-types/input-web-element.type";
import { WebElement } from "../element-types/web-element.type";

export class LoginPage extends BasePage {
    public readonly title: WebElement;
    public readonly emailInput: InputWebElement;
    public readonly passwordInput: InputWebElement;
    public readonly loginButton: WebElement;
    public readonly errorMessage: WebElement;
    private readonly rootElement: Locator;

    public constructor(readonly page: Page) {
        super(page);
        this.rootElement = this.page.locator('[data-testid="login-page"]');
        this.title = new WebElement(this.rootElement.locator('h1'), 'Title');
        this.emailInput = new InputWebElement(this.rootElement.locator('#email'), 'Email input field');
        this.passwordInput = new InputWebElement(this.rootElement.locator('#password'), 'Email input field');
        this.loginButton = new WebElement(this.rootElement.locator('[type="submit"]'), 'Login button');
        this.errorMessage = new WebElement(this.rootElement.locator('#error'), 'Error message');
    }

    public async validatePageIsLoaded(): Promise<void> {
        await Promise.all([
            this.title.validateElementVisible(),
            this.emailInput.validateElementVisible(),
            this.passwordInput.validateElementVisible(),
            this.loginButton.validateElementVisible(),
            this.errorMessage.validateElementNotVisible()
        ])
    }
}

File: LearnPlaywrightWithDotnet/src/MyTestAutomationFrameWork/PageObjects/LoginPage.cs
using Microsoft.Playwright;
using MyTestAutomationFramework.PageObjects;
using MyTestAutomationFrameWork.Core;
using static Microsoft.Playwright.Assertions;

public class LoginPage : BasePage
{
    // WebElements - Exposed for granular testing
    public WebElement EmailInput { get; }
    public WebElement PasswordInput { get; }
    public WebElement LoginButton { get; }
    public WebElement ErrorMessage { get; }
    public WebElement ForgotPasswordLink { get; }

    public LoginPage(IPage page) : base(page)
    {
        EmailInput = new WebElement(Page.Locator("#email"), "Email Input");
        PasswordInput = new WebElement(Page.Locator("#password"), "Password Input");
        LoginButton = new WebElement(Page.Locator("button[type='submit']"), "Login Button");
        ErrorMessage = new WebElement(Page.Locator(".error-message"), "Error Message");
        ForgotPasswordLink = new WebElement(Page.Locator("text=Forgot Password"), "Forgot Password Link");
    }

    // Business Methods - Encapsulate workflows
    public async Task LoginAsync(string email, string password)
    {
        await EmailInput.FillAsync(email);
        await PasswordInput.FillAsync(password);
        await LoginButton.ClickAsync();
    }

    public async Task LoginAndWaitForDashboardAsync(string email, string password)
    {
        await LoginAsync(email, password);
        await Expect(Page).ToHaveURLAsync(new Regex(".*/dashboard"));
    }

    public async Task VerifyLoginFailedAsync(string expectedError)
    {
        await ErrorMessage.ValidateElementVisibleAsync();
        await ErrorMessage.ContainTextAsync(expectedError);
    }

    public async Task<bool> IsLoginButtonEnabledAsync()
    {
        return await LoginButton.IsEnabledAsync();
    }
}